p2p/net/swarm/dial_worker.go

```diff
@@ -95,104 +113,117 @@ loop:
 			}
 
 			addrs, err := w.s.addrsForDial(req.ctx, w.peer)
+
 			if err != nil {
 				req.resch <- dialResponse{err: err}
 				continue loop
 			}
 
 			// at this point, len(addrs) > 0 or else it would be error from addrsForDial
-			// ranke them to process in order
-			addrs = w.rankAddrs(addrs)
+
+			// rank them to process in order
+			simConnect, _, _ := network.GetSimultaneousConnect(req.ctx)
+			addrDelays := w.rankAddrs(addrs, simConnect)
 
 			// create the pending request object
 			pr := &pendRequest{
 				req:   req,
 				err:   &DialError{Peer: w.peer},
 				addrs: make(map[ma.Multiaddr]struct{}),
 			}
-			for _, a := range addrs {
-				pr.addrs[a] = struct{}{}
+			for _, a := range addrDelays {
+				pr.addrs[a.Addr] = struct{}{}
 			}
 
 			// check if any of the addrs has been successfully dialed and accumulate
 			// errors from complete dials while collecting new addrs to dial/join
-			var todial []ma.Multiaddr
-			var tojoin []*addrDial
-
-			for _, a := range addrs {
-				ad, ok := w.pending[a]
+			tojoin := 0
+			newdials := 0
+			for _, nad := range addrDelays {
+				ad, ok := w.trackedDials[nad.Addr]
 				if !ok {
-					todial = append(todial, a)
+					w.dialQueue = append(w.dialQueue, nad)
+					w.trackedDials[nad.Addr] = &addrDial{
+						addr:     nad.Addr,
+						ctx:      req.ctx,
+						requests: []*pendRequest{pr},
+						delay:    nad.Delay,
+					}
+					newdials++
 					continue
 				}
 
-				if ad.conn != nil {
-					// dial to this addr was successful, complete the request
-					req.resch <- dialResponse{conn: ad.conn}
-					continue loop
-				}
-
+				// check if this dial has already errored.
+				// this dial couldn't have succeeded because bestAcceptableConnToPeer
+				// didn't return a connection
 				if ad.err != nil {
 					// dial to this addr errored, accumulate the error
-					pr.err.recordErr(a, ad.err)
-					delete(pr.addrs, a)
+					pr.err.recordErr(nad.Addr, ad.err)
```

Comment:
nit, use `ad.Addr`. It's the same thing, right?

Reply:
fixed.